derp/xdp/xdp.c

```diff
@@ -0,0 +1,585 @@
+
+struct stunxor6 {
+	__u8 unused;
+	__u8 family;
+	__be16 port;
+	__be32 addr[4];
+};
+
+#define STUN_BINDING_REQUEST 1
+
+#define STUN_MAGIC 0x2112a442
+
+#define STUN_ATTR_SW 0x8022
+
+#define STUN_ATTR_XOR_MAPPED_ADDR 0x0020
+
+#define STUN_BINDING_RESPONSE 0x0101
+
+#define STUN_MAGIC_FOR_PORT_XOR 0x2112
+
+#define MAX_UDP_LEN_IPV4 1480
+
+#define MAX_UDP_LEN_IPV6 1460
+
+#define IP_MF 0x2000
+#define IP_OFFSET 0x1fff
+
+static __always_inline __u16 csum_fold_flip(__u32 csum) {
+	__u32 sum;
+	sum = (csum >> 16) + (csum & 0xffff);
+	sum += (sum >> 16);
+	return ~sum;
+}
+
+// csum_const_size is an alternative to bpf_csum_diff. It's a verifier
+// workaround for when we are forced to use a constant max_size + bounds
+// checking. The alternative being passing a dynamic length to bpf_csum_diff
+// {from,to}_size arguments, which the verifier can't follow. For further info
+// see: https://github.com/iovisor/bcc/issues/2463#issuecomment-512503958
+static __always_inline __u16 csum_const_size(__u32 seed, void* from, void* data_end, int max_size) {
+	__u16 *buf = from;
+	for (int i = 0; i < max_size; i += 2) {
+		if ((void *)(buf + 1) > data_end) {
+			break;
+		}
+		seed += *buf;
+		buf++;
+	}
+	if ((void *)buf + 1 <= data_end) {
+		seed += *(__u8 *)buf;
+	}
+	return csum_fold_flip(seed);
+}
+
+static __always_inline __u32 pseudo_sum_ipv6(struct ipv6hdr* ip6, __u16 udp_len) {
+	__u32 pseudo = 0; // TODO(jwhited): __u64 for intermediate checksum values to reduce number of ops
+	for (int i = 0; i < 8; i ++) {
+		pseudo += ip6->saddr.in6_u.u6_addr16[i];
+		pseudo += ip6->daddr.in6_u.u6_addr16[i];
+	}
+	pseudo += bpf_htons(ip6->nexthdr);
+	pseudo += udp_len;
+	return pseudo;
+}
+
+static __always_inline __u32 pseudo_sum_ipv4(struct iphdr* ip, __u16 udp_len) {
+	__u32 pseudo = (__u16)ip->saddr;
+	pseudo += (__u16)(ip->saddr >> 16);
+	pseudo += (__u16)ip->daddr;
+	pseudo += (__u16)(ip->daddr >> 16);
+	pseudo += bpf_htons(ip->protocol);
+	pseudo += udp_len;
+	return pseudo;
+}
+
+struct packet_context {
+	enum counter_key_af af;
+	enum counter_key_prog_end prog_end;
+};
+
+static __always_inline int inc_counter(struct counters_key key, __u64 val) {
+	__u64 *counter = bpf_map_lookup_elem(&counters_map, &key);
+	if (!counter) {
+		return bpf_map_update_elem(&counters_map, &key, &val, BPF_ANY);
+	}
+	*counter += val;
+	return bpf_map_update_elem(&counters_map, &key, counter, BPF_ANY);
+}
+
+static __always_inline int handle_counters(struct xdp_md *ctx, int action, struct packet_context *pctx) {
+	void *data_end = (void *)(long)ctx->data_end;
+	void *data     = (void *)(long)ctx->data;
+
+	__u64 bytes = data_end - data;
+	enum counter_key_packets_bytes_action packets_pba = COUNTER_KEY_PACKETS_PASS_TOTAL;
+	enum counter_key_packets_bytes_action bytes_pba = COUNTER_KEY_BYTES_PASS_TOTAL;
+
+	switch (action) {
+		case XDP_ABORTED:
+			packets_pba = COUNTER_KEY_PACKETS_ABORTED_TOTAL;
+			bytes_pba = COUNTER_KEY_BYTES_ABORTED_TOTAL;
+			break;
+		case XDP_PASS:
```

Comment:
optional, minor: If you also make this `default:` then you don't need to initialize `packets_pba` and`bytes_pba` above.

Reply:
True, leaving as-is for (my subjective) readability